framework/include/cppmicroservices/BundleTracker.h

```diff
@@ -0,0 +1,266 @@
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+=============================================================================*/
+
+#ifndef CPPMICROSERVICES_BUNDLETRACKER_H
+#define CPPMICROSERVICES_BUNDLETRACKER_H
+
+#include <chrono>
+#include <map>
+#include <optional>
+#include <type_traits>
+
+#include "cppmicroservices/Bundle.h"
+#include "cppmicroservices/BundleEvent.h"
+#include "cppmicroservices/BundleTrackerCustomizer.h"
+
+namespace cppmicroservices {
+
+namespace detail {
+template<class T>
+class TrackedBundle;
+template<class T>
+class BundleTrackerPrivate;
+} // namespace detail
+
+class BundleContext;
+
+/**
+\defgroup gr_bundletracker BundleTracker
+
+\brief Groups BundleTracker related symbols.
+*/
+
+/**
+ * \ingroup MicroServices
+ * \ingroup gr_bundletracker
+ * 
+ * The <code>BundleTracker</code> class allows for keeping an accurate record of bundles and handling
+ * state changes, including bundles started before the tracker is opened.
+ *
+ * <p>
+ * A <code>BundleTracker</code> is constructed with a state mask and a <code>BundleTrackerCustomizer</code> object.
+ * The <code>BundleTracker</code> can select certain bundles to be tracked and trigger custom callbacks
+ * based on the provided <code>BundleTrackerCustomizer</code> or a <code>BundleTracker</code> subclass. Additionally,
+ * the template parameter allows for the use of custom tracked objects. Once the <code>BundleTracker</code>
+ * is opened, it will begin tracking all bundles that fall within the state mask.
+ *
+ * <p>
+ * Use the <code>getBundles</code> method to get the tracked <code>Bundle</code> objects, and <code>getObject</code> 
+ * to get the customized object.
+ *
+ * <p>
+ * The <code>BundleTracker</code> class is thread-safe. It does not call <code>BundleTrackerCustomizer</code> methods while
+ * holding any locks. Customizer implementations must be thread-safe.
+ *
+ * @tparam T The type of tracked object.
+ * @remarks This class is thread safe.
+ */
+template<class T = Bundle>
+class BundleTracker : protected BundleTrackerCustomizer<T>
+{
+public:
+  // The type of tracked object
+  using TrackedParamType =
+    typename BundleTrackerCustomizer<T>::TrackedParamType;
+
+  // The type of the tracking map
+  using TrackingMap = typename std::unordered_map<Bundle, TrackedParamType>;
+
+  // The type of the state mask
+  using BundleState = std::underlying_type_t<Bundle::State>;
+
+  /**
+   * Create a <code>StateType</code> stateMask for a BundleTracker
+   * 
+   * @param s0 the <code>Bundle::State</code>.
+   * @param s the subsequent <code>Bundle::States5</code>.
+   * 
+   * @return The state mask.
+   * 
+   */
+  template <typename S0, typename... S>
+  static constexpr typename BundleTracker<T>::BundleState CreateStateMask(S0 const &s0, S const &...s);
```

Comment:
It is true that `T` does not change the behavior of `BundleTracker<T>::CreateStateMask(...)`. However, it is not immediately obvious to me how to create a class method `BundleTracker::CreateStateMask(...)`, since class templates cannot share a name with a non-template classes.

Reply:
An update: The current commit is able to deduce the type of the `BundleTracker` *sometimes*. When a `BundleTracker` is constructed without a customizer it can perform the type deduction so the `<>` default template instation notation can be dropped:
```
cppms::BundleTracker bundleTracker(context, stateMask); // builds
```

In the gtest's however, this breaks down when we try to use mock customizers. I'm not sure why, but it is struggling with converting the third argument:
```
class MockCustomizer : BundleTrackerCustomizer<>
{ /* ... */ }

MockCustomizer customizer;
cppms::BundleTracker bundleTracker(context, stateMask, customizer); // fails to build
```

Additionally, `CreateStateMask` fails to deduce the type of `BundleTracker`. I think this is because the parameters for that method are never the template type. Unfortunately, one can't simply add an additional declaration for a nontemplate `BundleTracker` (even though `CreateStateMask` really does not depend on the type). However, another option that could certainly work would be to simply change `CreateStateMask` from a static method to a first-class function in the `cppmicroservices` namespace, while keeping it in the `BundleTracker` header. This could set a bad precedent of namespace pollution, but would completely separate the state mask generation from the templated `BundleTracker`.

It is not obvious to me how to proceed in further enabling these type deductions, but it feels like there must be a way to fix the edge cases.